File: cpp/examples/activity_feed/activity_item/activity_item.cpp
#include "activity_item.hpp"

#include "activity_item_ui.hpp"
#include "ui/generated/GeneratedWindowRuntime.hpp"

#include <algorithm>
#include <sstream>
#include <string>
#include <string_view>
#include <vector>

namespace activity_feed_example {

namespace {

[[nodiscard]] std::vector<std::string_view> splitLines(std::string_view text) {
    std::vector<std::string_view> lines;
    std::size_t offset = 0;
    while (offset <= text.size()) {
        const std::size_t newline = text.find('\n', offset);
        if (newline == std::string_view::npos) {
            lines.push_back(text.substr(offset));
            break;
        }
        lines.push_back(text.substr(offset, newline - offset));
        offset = newline + 1;
    }
    if (lines.empty()) {
        lines.emplace_back();
    }
    return lines;
}

[[nodiscard]] int wrappedLineCount(std::string_view text, int width) {
    const int safeWidth = std::max(1, width);
    int count = 0;
    for (std::string_view line : splitLines(text)) {
        count += std::max(1, static_cast<int>((line.size() + static_cast<std::size_t>(safeWidth) - 1) /
                                              static_cast<std::size_t>(safeWidth)));
    }
    return std::max(1, count);
}

template <typename GeneratedView>
[[nodiscard]] int generatedSourceHeight(const GeneratedView& view) {
    int height = 0;
    for (const auto& entry : view.generatedLayout()) {
        height = std::max(height, entry.sourceCell.row + entry.sourceCell.height);
    }
    return std::max(1, height);
}

template <typename GeneratedView>
[[nodiscard]] int compactContentHeight(const GeneratedView& view, int messageRows) {
    int height = 1;
    for (const auto& entry : view.generatedLayout()) {
        if (entry.name.empty()) {
            continue;
        }
        const int entryHeight = entry.name == "message"
            ? std::max(1, messageRows)
            : std::max(1, entry.charsSize.height);
        height = std::max(height, entry.relative.row + entryHeight);
    }
    return height;
}

[[nodiscard]] ui::Color markerColor(std::string_view eventType) {
    if (eventType == "Task") {
        return ui::Color{"#a78bfa"};
    }
    if (eventType == "Warning") {
        return ui::Color{"#facc15"};
    }
    if (eventType == "Deploy") {
        return ui::Color{"#22c55e"};
    }
    if (eventType == "Note") {
        return ui::Color{"#fb7185"};
    }
    return ui::Color{"#38bdf8"};
}

void applyLeftMarker(ui::RenderedContent& rendered, const ui::Color& color) {
    for (ui::RenderedRow& row : rendered) {
        if (row.empty()) {
            continue;
        }
        ui::TerminalCell& cell = row.front();
        cell.text = "▐";
        cell.foreground = color;
    }
}

}  // namespace

ui::RenderedContent renderActivityItem(const ActivityItemData& item, int width) {
    ActivityItemUI view;
    view.timestamp->setText(item.showTimestamp ? item.timestamp : "");
    view.event_type->setText(item.eventType);
    view.message->setText(item.message);

    const int sourceHeight = generatedSourceHeight(view);
    (void)ui::renderGeneratedWindowContent(view, ui::Size{std::max(1, width), sourceHeight});
    const int messageRows = wrappedLineCount(item.message, std::max(1, view.message->frame().width));
    ui::RenderedContent rendered = ui::renderGeneratedWindowContent(
        view,
        ui::Size{
            std::max(1, width),
            compactContentHeight(view, messageRows),
        });
    applyLeftMarker(rendered, markerColor(item.eventType));
    return rendered;
}

}  // namespace activity_feed_example

File: go/examples/activity_feed/activity_item/activity_item.go
package activity_item

import (
	"strings"

	uimd "uimd"
)

const (
	activityItemMinimumExtent = 1
	ActivityItemDefaultHeight = activityItemMinimumExtent
)

type ActivityItemData struct {
	Timestamp     string
	EventType     string
	Message       string
	ShowTimestamp bool
}

type ActivityItemElement struct {
	name  string
	frame uimd.Rect
	data  ActivityItemData
}

func NewActivityItemElement(name string, data ActivityItemData) *ActivityItemElement {
	return &ActivityItemElement{
		name:  name,
		frame: uimd.Rect{Width: 1, Height: ActivityItemDefaultHeight},
		data:  data,
	}
}

func (element *ActivityItemElement) ElementName() string {
	return element.name
}

func (element *ActivityItemElement) ElementType() string {
	return "activity_item"
}

func (element *ActivityItemElement) IsFocusable() bool {
	return false
}

func (element *ActivityItemElement) IsEnabled() bool {
	return true
}

func (element *ActivityItemElement) SetFrame(frame uimd.Rect) {
	element.frame = frame
}

func (element *ActivityItemElement) ElementFrame() uimd.Rect {
	return element.frame
}

func (element *ActivityItemElement) HandleKey(string) bool {
	return false
}

func (element *ActivityItemElement) SetText(string) {
}

func (element *ActivityItemElement) Render(size uimd.Size, state uimd.ElementRenderState) [][]uimd.TerminalCell {
	view := NewActivityItemUI()
	configureActivityItemView(view, element.data)
	renderHeight := element.PreferredHeight(size.Width)
	renderSize := uimd.Size{Width: maxPositive(size.Width), Height: renderHeight}
	buffer := uimd.RenderGeneratedWindowContent(view.GeneratedWindowBase, renderSize, -1, nil, false)
	rows := bufferCells(buffer, renderSize)
	color := uimd.NewColor(markerColor(element.data.EventType))
	for row := range rows {
		if len(rows[row]) == 0 {
			continue
		}
		rows[row][0].Text = "▐"
		rows[row][0].Foreground = color
	}
	return rows
}

func (element *ActivityItemElement) PreferredHeight(width int) int {
	view := NewActivityItemUI()
	configureActivityItemView(view, element.data)
	sourceHeight := generatedSourceHeight(view.GeneratedWindowBase)
	_ = uimd.RenderGeneratedWindowContent(
		view.GeneratedWindowBase,
		uimd.Size{Width: maxPositive(width), Height: sourceHeight},
		-1,
		nil,
		false)
	messageRows := WrappedLineCount(element.data.Message, maxPositive(view.Message.ElementFrame().Width))
	return compactContentHeight(view.GeneratedWindowBase, messageRows)
}

func RenderActivityItem(data ActivityItemData, width int) *ActivityItemElement {
	element := NewActivityItemElement("activity_item", data)
	element.SetFrame(uimd.Rect{Width: width, Height: element.PreferredHeight(width)})
	return element
}

func configureActivityItemView(view *ActivityItemUI, data ActivityItemData) {
	if data.ShowTimestamp {
		view.Timestamp.SetText(data.Timestamp)
	} else {
		view.Timestamp.SetText("")
	}
	view.EventType.SetText(data.EventType)
	view.Message.SetText(data.Message)
}

func generatedSourceHeight(window *uimd.GeneratedWindowBase) int {
	height := activityItemMinimumExtent
	for _, entry := range window.GeneratedLayout() {
		height = maxValue(height, entry.CellRow+entry.CellCharsHeight)
	}
	return height
}

func compactContentHeight(window *uimd.GeneratedWindowBase, messageRows int) int {
	height := activityItemMinimumExtent
	for _, entry := range window.GeneratedLayout() {
		if entry.Content == "" {
			continue
		}
		entryHeight := entry.CharsHeight
		if entry.Content == "message" {
			entryHeight = maxPositive(messageRows)
		}
		height = maxValue(height, entry.Row+maxPositive(entryHeight))
	}
	return height
}

func maxPositive(value int) int {
	if value < activityItemMinimumExtent {
		return activityItemMinimumExtent
	}
	return value
}

func maxValue(first int, second int) int {
	if first > second {
		return first
	}
	return second
}

func markerColor(eventType string) string {
	switch eventType {
	case "Task":
		return "#a78bfa"
	case "Warning":
		return "#facc15"
	case "Deploy":
		return "#22c55e"
	case "Note":
		return "#fb7185"
	default:
		return "#38bdf8"
	}
}

func bufferCells(buffer *uimd.TerminalBuffer, size uimd.Size) [][]uimd.TerminalCell {
	rows := make([][]uimd.TerminalCell, size.Height)
	for row := 0; row < size.Height; row++ {
		rows[row] = make([]uimd.TerminalCell, size.Width)
		for col := 0; col < size.Width; col++ {
			rows[row][col] = buffer.Cell(row, col)
		}
	}
	return rows
}

func WrappedLineCount(text string, width int) int {
	if width < 1 {
		width = 1
	}
	count := 0
	for _, line := range strings.Split(text, "\n") {
		lineWidth := len([]rune(line))
		rows := (lineWidth + width - 1) / width
		if rows < 1 {
			rows = 1
		}
		count += rows
	}
	if count < 1 {
		return 1
	}
	return count
}
